shopinvader/components/access_info.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2019 Camptocamp (http://www.camptocamp.com).
+# @author Simone Orsi <simone.orsi@camptocamp.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.addons.component.core import Component
+
+
+class PartnerAccess(Component):
+    """Define access rules to partner from client side."""
+
+    _name = "shopinvader.partner.access"
+    _inherit = "base.shopinvader.component"
+    _usage = "access.info"
+    _apply_on = "res.partner"
+
+    @property
+    def service_work(self):
+        return getattr(self.work, "service_work", None)
+
+    @property
+    def partner(self):
+        return getattr(self.work, "partner", None)
+
+    @property
+    def partner_user(self):
+        return getattr(self.work, "partner_user", self.partner)
+
+    def is_main_partner(self):
+        return self.partner == self.partner_user
+
+    def is_owner(self, partner_id):
+        return partner_id == self.partner_user.id
+
+    def for_profile(self, partner_id):
+        info = {"read": True, "update": True, "delete": False}
+        if not self.is_main_partner() and partner_id != self.partner_user.id:
+            info["update"] = False
+        return info
+
+    def for_address(self, address_id):
+        info = {"read": True, "update": True, "delete": True}
+        if self.partner_user is not None:
+            if not self.is_main_partner():
+                if not self.is_owner(address_id):
+                    info.update(
+                        {"read": True, "update": False, "delete": False}
+                    )
+                else:
+                    # only main partner can delete your address
+                    info["delete"] = False
+        return info
+
+    def permission(self):
+        if self.partner is None:
+            return {"address": {}, "purchase": {}}
+        return {
+            # scope: permissions
+            "address": {
+                # can create addresses only if profile partner is enabled
+                "create": self.partner.shopinvader_enabled
+            },
+            "purchase": {
```

Comment:
I would use the name of the service and method as keys. This will allow us the same info to enforce these security rules at the service level. A simple override of dispatch https://github.com/OCA/rest-framework/blob/12.0/base_rest/components/service.py#L174 into BaseShopinvaderService method https://github.com/shopinvader/odoo-shopinvader/blob/12.0/shopinvader/services/service.py could do the job...

```python
    @property
    def access_info(self):
        with self.shopinvader_backend.work_on(
            "res.partner",
            partner=self.partner,
            partner_user=self.partner_user,
            service_work=self.work,
        ) as work:
            return work.component(usage="access.info")

    def _check_permission(method_name, _id=None, params=None):
        permissions = self.access_info.permission
        allowed_methods = permissions.get(self._usage, {})
        is_method_allowed = allowed_methods.get(method_name, False)
        if not is_method_allowed:
            raise AccessErrror()

    def dispatch(self, method_name, _id=None, params=None):
        self._check_permission(method_name, _id=_id, params=params)
        return super().dispatch(method_name, _id=_id, params=params)
```
```